### PR TITLE
Add argument names to docstrings in cocotb.simulator

### PR DESCRIPTION
**Pull request by eric-wieser**
Note that `/` is special syntax meaning "positional-only".

C docstrings of the form
```
the_func(the_args)
--

the docstring
```
and
```
the_method($self, the_args)
--

the docstring
```
automatically populate the `__text_signature__` attribute, which is read by `inspect.signature` (and therefore `help()`, `?`, and ipython autocomplete).

---

First commit is #1837, will rebase once that is merged.

It seems the latest release of sphinx does not understand `__text_signature__`, but master (https://github.com/sphinx-doc/sphinx/pull/7623) does.

Either way, the docs at https://cocotb--1838.org.readthedocs.build/en/1838/library_reference.html#cocotb.simulator.gpi_cb_hdl should make it clear that the `--` syntax is being consumed by python itself, as it does not appear verbatim in the docs.

### CODE REVIEW

**Comment by eric-wieser** [comment]
As a follow-up, we could consider pinning to a random sphinx commit in master so that these signatures appear.

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 1]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
 
 static PyMethodDef gpi_sim_hdl_methods[] = {
     {"get_signal_val_long",
-        (PyCFunction)get_signal_val_long, METH_NOARGS,
-        "Get the value of a signal as a long"},
+        (PyCFunction)get_signal_val_long, METH_NOARGS, PyDoc_STR(
+            "get_signal_val_long($self)\n"
+            "--\n\n"
+            "Get the value of a signal as a long"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:929]:
```suggestion
            "Get the value of a signal as an int"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 2]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"get_signal_val_str",
-        (PyCFunction)get_signal_val_str, METH_NOARGS,
-        "Get the value of a signal as an ASCII string"},
+        (PyCFunction)get_signal_val_str, METH_NOARGS, PyDoc_STR(
+            "get_signal_val_str($self)\n"
+            "--\n\n"
+            "Get the value of a signal as an ASCII string"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:936]:
```suggestion
            "Get the value of a signal as an string"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 3]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"get_signal_val_binstr",
-        (PyCFunction)get_signal_val_binstr, METH_NOARGS,
-        "Get the value of a signal as a binary string"},
+        (PyCFunction)get_signal_val_binstr, METH_NOARGS, PyDoc_STR(
+            "get_signal_val_binstr($self)\n"
+            "--\n\n"
+            "Get the value of a signal as a binary string"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:943]:
```suggestion
            "Get the value of a logic vector signal as a string ('0', '1', 'X', etc.), one element per character"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 4]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"get_signal_val_real",
-        (PyCFunction)get_signal_val_real, METH_NOARGS,
-        "Get the value of a signal as a double precision float"},
+        (PyCFunction)get_signal_val_real, METH_NOARGS, PyDoc_STR(
+            "get_signal_val_real($self)\n"
+            "--\n\n"
+            "Get the value of a signal as a double precision float"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:950]:
```suggestion
            "Get the value of a signal as a float"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 5]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"set_signal_val_long",
-        (PyCFunction)set_signal_val_long, METH_VARARGS,
-        "Set the value of a signal using a long"},
+        (PyCFunction)set_signal_val_long, METH_VARARGS, PyDoc_STR(
+            "set_signal_val_long($self, action, value, /)\n"
+            "--\n\n"
+            "Set the value of a signal using a long"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:957]:
```suggestion
            "Set the value of a signal using an int"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 6]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"set_signal_val_str",
-        (PyCFunction)set_signal_val_str, METH_VARARGS,
-        "Set the value of a signal using an NUL-terminated 8-bit string"},
+        (PyCFunction)set_signal_val_str, METH_VARARGS, PyDoc_STR(
+            "set_signal_val_str($self, action, value, /)\n"
+            "--\n\n"
+            "Set the value of a signal using an NUL-terminated 8-bit string"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:964]:
```suggestion
            "Set the value of a signal using a user-encoded string as bytes"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 7]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"set_signal_val_binstr",
-        (PyCFunction)set_signal_val_binstr, METH_VARARGS,
-        "Set the value of a signal using a string with a character per bit"},
+        (PyCFunction)set_signal_val_binstr, METH_VARARGS, PyDoc_STR(
+            "set_signal_val_binstr($self, action, value, /)\n"
+            "--\n\n"
+            "Set the value of a signal using a string with a character per bit"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:971]:
```suggestion
            "Set the value of a logic vector signal as a string ('0', '1', 'X', etc.), one element per character"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 8]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"set_signal_val_real",
-        (PyCFunction)set_signal_val_real, METH_VARARGS,
-        "Set the value of a signal using a double precision float"},
+        (PyCFunction)set_signal_val_real, METH_VARARGS, PyDoc_STR(
+            "set_signal_val_real($self, action, value, /)\n"
+            "--\n\n"
+            "Set the value of a signal using a double precision float"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:978]:
```suggestion
            "Set the value of a signal using a float"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 9]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+            "Get the file that sources the object's definition"
+        )
+    },
     {"get_handle_by_name",
-        (PyCFunction)get_handle_by_name, METH_VARARGS,
-        "Get handle of a named object"},
+        (PyCFunction)get_handle_by_name, METH_VARARGS, PyDoc_STR(
+            "get_handle_by_name($self, name, /)\n"
+            "--\n\n"
+            "Get handle of a named object"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:999]:
```suggestion
            "Get handle to a child object by name"
```
This now works by getting a handle to a sub-object only, right?

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 10]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"get_handle_by_index",
-        (PyCFunction)get_handle_by_index, METH_VARARGS,
-        "Get handle of a object at an index in a parent"},
+        (PyCFunction)get_handle_by_index, METH_VARARGS, PyDoc_STR(
+            "get_handle_by_index($self, index, /)\n"
+            "--\n\n"
+            "Get handle of a object at an index in a parent"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:1006]:
```suggestion
            "Get handle to a child object by index"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 11]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+            "Get the name of an object as a string"
+        )
+    },
     {"get_type_string",
-        (PyCFunction)get_type_string, METH_NOARGS,
-        "Get the type of an object as a string"},
+        (PyCFunction)get_type_string, METH_NOARGS, PyDoc_STR(
+            "get_type_string($self)\n"
+            "--\n\n"
+            "Get the type of an object as a string"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:1020]:
```suggestion
            "Get the GPI enumeration type of an object as a string"
```

This got me the first time I used it. I thought it was going to be useful...

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 12]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"get_type",
-        (PyCFunction)get_type, METH_NOARGS,
-        "Get the type of an object, mapped to a GPI enumeration"},
+        (PyCFunction)get_type, METH_NOARGS, PyDoc_STR(
+            "get_type($self)\n"
+            "--\n\n"
+            "Get the type of an object, mapped to a GPI enumeration"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:1027]:
```suggestion
            "Get the GPI enumeration type of an object"
```

File: cocotb/share/lib/simulator/simulatormodule.cpp [hunk 13]
```diff
@@ -923,62 +923,138 @@ PyMODINIT_FUNC PyInit_simulator(void)
...
+        )
+    },
     {"get_const",
-        (PyCFunction)get_const, METH_NOARGS,
-        "Get a flag indicating whether the object is a constant"},
+        (PyCFunction)get_const, METH_NOARGS, PyDoc_STR(
+            "get_const($self)\n"
+            "--\n\n"
+            "Get a flag indicating whether the object is a constant"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.cpp:1034]:
```suggestion
            "Return ``True`` if the object is constant"
```

File: cocotb/share/lib/simulator/simulatormodule.h [hunk 14]
```diff
@@ -70,20 +70,72 @@ static PyObject *get_simulator_version(PyObject *self, PyObject *args);
...
+    )},
+    {"register_rwsynch_callback", register_rwsynch_callback, METH_VARARGS, PyDoc_STR(
+        "register_rwsynch_callback(func, /, *args)\n"
+        "--\n\n"
+        "Register a callback for the read-write section"
+    )},
+    {"stop_simulator", stop_simulator, METH_VARARGS, PyDoc_STR(
+        "stop_simulator()\n"
+        "--\n\n"
+        "Instruct the attached simulator to stop"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.h:111]:
```suggestion
        "Instruct the attached simulator to stop. Users should not call this function."
```

File: cocotb/share/lib/simulator/simulatormodule.h [hunk 15]
```diff
@@ -70,20 +70,72 @@ static PyObject *get_simulator_version(PyObject *self, PyObject *args);
...
+    )},
+    {"get_precision", get_precision, METH_NOARGS, PyDoc_STR(
+        "get_precision()\n"
+        "--\n\n"
+        "Get the precision of the simulator"
+    )},
+    {"get_simulator_product", get_simulator_product, METH_NOARGS, PyDoc_STR(
+        "get_simulator_product()\n"
+        "--\n\n"
+        "Simulator product information"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.h:132]:
```suggestion
        "Get the simulator's product string"
```

File: cocotb/share/lib/simulator/simulatormodule.h [hunk 16]
```diff
@@ -70,20 +70,72 @@ static PyObject *get_simulator_version(PyObject *self, PyObject *args);
...
+    )},
+    {"get_simulator_version", get_simulator_version, METH_NOARGS, PyDoc_STR(
+        "get_simulator_version()\n"
+        "--\n\n"
+        "Simulator product version information"
```
Comment (ktbarrett) [inline cocotb/share/lib/simulator/simulatormodule.h:137]:
```suggestion
        "Get the simulator's product version string"
```

**Comment by eric-wieser** [comment]
> The docstrings could be better, and here should reference the Python types they expect and not the C types.

My intent of this PR was not to touch the wording of the docstrings, but just to make `inspect.signature` work, which it does.

If you run `help(cocotb.simulator)` in an ipython shell, this patch results in all the signatures appearing.

**Comment by eric-wieser** [comment]
Thanks! Please do follow up with a PR to improve these docstrings :)

File: cocotb/share/lib/simulator/simulatormodule.cpp (hunk 2 above)
Comment (cmarqu) [reply cocotb/share/lib/simulator/simulatormodule.cpp:936]: Typo in the suggestion: "Get the value of a signal as a string"

**Pull request merged**